File: core/aggregators/rss.py
"""RSS aggregator base class."""

from datetime import datetime
from datetime import timezone as dt_timezone
from email.utils import parsedate_to_datetime
from typing import Any, Dict, List, Optional
from urllib.parse import urlparse

from django.utils import timezone

from .base import BaseAggregator
from .utils import discover_feed_url, parse_rss_feed


class RssAggregator(BaseAggregator):
    """Base class for RSS-based aggregators."""

    def __init__(self, feed):
        super().__init__(feed)

    def aggregate(self) -> List[Dict[str, Any]]:
        """Implement template method pattern flow."""
        self.validate()
        limit = self.get_current_run_limit()
        if limit == 0:
            return []
        source_data = self.fetch_source_data(limit)
        articles = self.parse_to_raw_articles(source_data)
        articles = self.filter_articles(articles)
        articles = self.enrich_articles(articles)
        articles = self.finalize_articles(articles)
        return articles

    def fetch_source_data(self, limit: Optional[int] = None) -> Dict[str, Any]:
        """Fetch RSS feed data, following a homepage to its advertised feed.

        ``parse_rss_feed`` raises ``ValueError`` both when the identifier is not
        a feed and when it yields zero entries. Either way, if the identifier
        looks like a page URL, try the feed the page advertises. Best-effort: an
        identifier with no discoverable feed re-raises the original error, so the
        outcome stays the existing "no entries" one rather than a new error class.
        """
        self.logger.info(f"Fetching RSS feed: {self.identifier}")
        try:
            return parse_rss_feed(self.identifier)
        except ValueError:
            parsed = urlparse(self.identifier or "")
            if not parsed.scheme or not parsed.netloc:
                raise

            discovered = discover_feed_url(self.identifier)
            if not discovered or discovered == self.identifier:
                raise

            self.logger.info(f"Discovered feed for {self.identifier}: {discovered}")
            return parse_rss_feed(discovered)

    def parse_to_raw_articles(self, source_data: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Parse RSS feed items to article dictionaries."""
        articles = []
        entries = source_data.get("entries", [])
        limit = self.get_current_run_limit()

        for entry in entries[:limit]:
            article = {
                "name": entry.get("title", ""),
                "identifier": entry.get("link", ""),
                "raw_content": "",  # To be filled by enrich_articles
                "content": entry.get("summary", ""),
                "date": self._parse_date(entry.get("published")),
                "author": entry.get("author", ""),
                "icon": None,
            }
            articles.append(article)

        return articles

    def _parse_date(self, date_str: Optional[str]) -> datetime:
        """Parse an RSS date string into an aware datetime.

        ``parsedate_to_datetime`` returns a naive datetime both when the
        source has no timezone at all and for the RFC 5322 "-0000" offset
        ("UTC, local zone unknown"). Either way there is no basis for
        assuming the server's local ``TIME_ZONE`` -- attach UTC instead, so
        the instant is not silently shifted by whatever zone the server
        happens to run in.
        """
        if not date_str:
            return timezone.now()
        try:
            parsed = parsedate_to_datetime(date_str)
        except Exception:
            return timezone.now()
        if timezone.is_naive(parsed):
            return parsed.replace(tzinfo=dt_timezone.utc)
        return parsed

File: src/lib/aggregators/rss.ts
import { BaseAggregator, RawArticle } from "./base";
import { discoverFeedUrl, parseRssFeed, ParsedFeed, unescapeEntities } from "./rss-parser";

export class RssAggregator extends BaseAggregator {
  static unescapeEntities = unescapeEntities;

  protected parseDate(dateStr?: string | null): Date {
    if (!dateStr) return new Date();
    const parsed = new Date(dateStr);
    if (isNaN(parsed.getTime())) {
      return new Date();
    }
    return parsed;
  }

  async fetchSourceData(_limit?: number): Promise<ParsedFeed> {
    try {
      return await parseRssFeed(this.identifier);
    } catch (err) {
      let isUrl = false;
      try {
        const u = new URL(this.identifier);
        isUrl = Boolean(u.protocol && u.host);
      } catch {
        isUrl = false;
      }

      if (!isUrl) {
        throw err;
      }

      const discovered = await discoverFeedUrl(this.identifier);
      if (!discovered || discovered === this.identifier) {
        throw err;
      }

      return await parseRssFeed(discovered);
    }
  }

  /**
   * A plain RSS/"Feed Content" feed never fetches a full page -- the entry's
   * own `summary` *is* the article's content, both on first aggregation
   * (`parseToRawArticles()` below) and here on reload. So "fetch from
   * source" for this aggregator means re-fetching the feed itself and
   * finding this article's entry again by its link, not fetching `url` as a
   * page. Returns "" (reload.ts's existing "could not be reloaded" branch)
   * when the feed can no longer be reached or no longer lists this entry --
   * e.g. it aged out of the feed's own window.
   */
  async fetchArticleContent(url: string): Promise<string> {
    try {
      const feed = await this.fetchSourceData();
      const entry = feed.entries.find((e) => e.link === url);
      // Unescaped exactly as parseToRawArticles() below does it, so reload and
      // a fresh aggregation run report the same title for the same entry. It is
      // what keeps reload from feeding a previous AI run's title back into the
      // AI stage -- see `noteSourceTitle()` in ./base.
      this.noteSourceTitle(entry ? unescapeEntities(entry.title || "") : null);
      return entry?.summary || "";
    } catch {
      return "";
    }
  }

  async parseToRawArticles(sourceData: unknown): Promise<RawArticle[]> {
    const feed = sourceData as ParsedFeed;
    const entries = feed?.entries || [];
    const articles: RawArticle[] = [];
    const limit = this.getCurrentRunLimit();

    for (const entry of entries.slice(0, limit > 0 ? limit : entries.length)) {
      articles.push({
        name: unescapeEntities(entry.title || ""),
        identifier: entry.link || "",
        raw_content: "",
        content: entry.summary || "",
        date: this.parseDate(entry.published),
        author: unescapeEntities(entry.author || ""),
        icon: null,
      });
    }

    return articles;
  }
}
